src/Data/Vector/Linear.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | This module defines vectors of known length which can hold linear values
+--
+-- Having a known length matters with linear types, because many common vector
+-- operations (like zip) are not total with linear types.
+--
+-- A much more expensive library of vectors of known size (including matrices
+-- and tensors of all dimensions) is the [@linear@ library on
+-- Hackage](https://hackage.haskell.org/package/linear) (that's /linear/ in the
+-- sense of [linear algebra](https://en.wikipedia.org/wiki/Linear_algebra),
+-- rather than linear types).
+
+module Data.Vector.Linear
+  ( V
+  , Elim
+  , elim
+  ) where
+
+import qualified Control.Monad.Linear as Control
+import qualified Data.Functor.Linear as Data
+import Data.Proxy
+import Data.Type.Equality
+import Data.Vector (Vector)
+import qualified Data.Vector as Vector
+import GHC.Exts (Constraint, proxy#)
+import GHC.TypeLits
+import Prelude
+  ( Eq
+  , Ord
+  , Int
+  , Bool(..)
+  , Either(..)
+  , Maybe(..)
+  , fromIntegral
+  , error
+  , (-))
+import qualified Prelude as Prelude
+import Prelude.Linear.Internal.Simple
+import qualified Unsafe.Linear as Unsafe
+
+newtype V (n :: Nat) (a :: *) = V (Vector a)
+  deriving (Eq, Ord, Prelude.Functor)
+  -- Using vector rather than, say, 'Array' (or directly 'Array#') because it
+  -- offers many convenience function. Since all these unsafeCoerces probably
+  -- kill the fusion rules, it may be worth it going lower level since I
+  -- probably have to write my own fusion anyway. Therefore, starting from
+  -- Arrays at the moment.
```

Comment:
Should this say "starting from Vectors at the moment"?

Reply:
Yes.